File: java-suite/src/main/java/de/fraunhofer/iem/generics/upperBounds/inGenericClass/CustomPair.java
package de.fraunhofer.iem.generics.upperBounds.inGenericClass;

/**
 * Demonstrates the upper bounds in generics class in Java
 *
 * @author Ranjith Krishnamurthy
 */
public class CustomPair<A extends Number, B extends CharSequence & Comparable<B>> {
    private A first;
    private B second;

    public CustomPair(A first, B second) {
        this.first = first;
        this.second = second;
    }

    public A getFirst() {
        return first;
    }

    public B getSecond() {
        return second;
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/generics/upperBounds/inGenericClass/CustomPair.kt
package de.fraunhofer.iem.generics.upperBounds.inGenericClass

/**
 * Demonstrates the upperbound in generics class in Kotlin
 *
 * @author Ranjith Krishnamurthy
 */
data class CustomPair<A, B>(val first: A, val second: B) where A : Number,
                                                               B : CharSequence,
                                                               B : Comparable<B>